Array out of bounds issue causing panic

## Bug Report

<!--
If you have usage questions, please try the [Discussions](https://github.com/kubearmor/KubeArmor/discussions) or [KubeArmor Slack](https://kubearmor.slack.com/) first.
Please check the [Kubearmor issue](https://github.com/kubearmor/KubeArmor/issues) list to check if there is an issue already in the context.
-->

**General Information**

- Azure AKS
- 5.15.0-1091-azure
- 1.32.5
- CLI installed 1.44
- Target pod - N/A

We've been running KubeArmor for around 2 weeks, and last night we had a panic in a pod in the daemon set which seems to be internal.

Panic: 
```
E0815 00:30:47.985743       1 panic.go:115] "Observed a panic" panic="runtime error: index out of range [39] with length 39" panicGoValue="runtime.boundsError{x:39, y:39, signed:true, code:0x0}" stacktrace=<
        goroutine 475 [running]:
        k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2a59a20, 0xc005956cf0}, {0x2512520, 0xc0038d6048})
                /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:132 +0xbc
        k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x2a59a20, 0xc005956cc0}, {0x2512520, 0xc0038d6048}, {0x0, 0x0, 0x0?})
                /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:107 +0x116
        k8s.io/apimachinery/pkg/util/runtime.HandleCrashWithLogger({{0x2a60c48?, 0xc00004f500?}, 0xc0057d8640?}, {0x0, 0x0, 0x0})
                /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:91 +0x115
        panic({0x2512520?, 0xc0038d6048?})
                /usr/local/go/src/runtime/panic.go:792 +0x132
        github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).UpdateSecurityPolicy(_, {_, _}, {_, _}, {0xc004e00b70, {{0x0, {0x0, 0x0, 0x0}, ...}, ...}})
                /usr/src/KubeArmor/KubeArmor/core/kubeUpdate.go:1074 +0x2a87
        github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).WatchSecurityPolicies.func2({0x2674d40?, 0xc0057d9f70?}, {0x2674d40?, 0xc005b43c08?})
                /usr/src/KubeArmor/KubeArmor/core/kubeUpdate.go:1775 +0x4d8
        k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
                /go/pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/controller.go:264
        k8s.io/client-go/tools/cache.(*processorListener).run.func1(0xc0057d9f5f, 0xc005c3c780, {0x23e0440?, 0xc005790780?})
                /go/pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/shared_informer.go:1075 +0x17d
        k8s.io/client-go/tools/cache.(*processorListener).run(0xc005c3c780)
                /go/pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/shared_informer.go:1087 +0x3c
        k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
                /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/wait/wait.go:72 +0x4c
        created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 392
                /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/wait/wait.go:70 +0x73
 >
panic: runtime error: index out of range [39] with length 39 [recovered]
        panic: runtime error: index out of range [39] with length 39

goroutine 475 [running]:
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x2a59a20, 0xc005956cc0}, {0x2512520, 0xc0038d6048}, {0x0, 0x0, 0x0?})
        /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:114 +0x1a9
k8s.io/apimachinery/pkg/util/runtime.HandleCrashWithLogger({{0x2a60c48?, 0xc00004f500?}, 0xc0057d8640?}, {0x0, 0x0, 0x0})
        /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/runtime/runtime.go:91 +0x115
panic({0x2512520?, 0xc0038d6048?})
        /usr/local/go/src/runtime/panic.go:792 +0x132
github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).UpdateSecurityPolicy(_, {_, _}, {_, _}, {0xc004e00b70, {{0x0, {0x0, 0x0, 0x0}, ...}, ...}})
        /usr/src/KubeArmor/KubeArmor/core/kubeUpdate.go:1074 +0x2a87
github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).WatchSecurityPolicies.func2({0x2674d40?, 0xc0057d9f70?}, {0x2674d40?, 0xc005b43c08?})
        /usr/src/KubeArmor/KubeArmor/core/kubeUpdate.go:1775 +0x4d8
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
        /go/pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/controller.go:264
k8s.io/client-go/tools/cache.(*processorListener).run.func1(0xc0057d9f5f, 0xc005c3c780, {0x23e0440?, 0xc005790780?})
        /go/pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/shared_informer.go:1075 +0x17d
k8s.io/client-go/tools/cache.(*processorListener).run(0xc005c3c780)
        /go/pkg/mod/k8s.io/client-go@v0.33.1/tools/cache/shared_informer.go:1087 +0x3c
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
        /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/wait/wait.go:72 +0x4c
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 392
        /go/pkg/mod/k8s.io/apimachinery@v0.33.1/pkg/util/wait/wait.go:70 +0x73
```

Investigating we don't think this is our issue with the deployment, but looking at the panic stack trace we see that [this section](https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/core/kubeUpdate.go#L1092C1-L1099C29) reads the length of an array within a read lock, and then unlocks, but then later iterates on that length and uses it to access the array. I believe another goroutine is probably removing an element from that array while the read lock is unlocked, and this means that the final access (in our panic, 39) is out of range.

**To Reproduce**

Unsure, This seems to be an internal issue with the service. The pod was running for around 32 hours before the issue.

**Expected behavior**

Kubearmor pod in daemon set should not panic.

**Screenshots**

N/A

--- a/KubeArmor/core/containerdHandler.go
+++ b/KubeArmor/core/containerdHandler.go
@@ -527,6 +527,7 @@ func (dm *KubeArmorDaemon) UpdateContainerdContainer(ctx context.Context, contai
 
 		// delete endpoint if no security rules and containers
 		if !dm.K8sEnabled {
+			dm.EndPointsLock.Lock()
 			idx := 0
 			endpointsLength := len(dm.EndPoints)
 			for idx < endpointsLength {
@@ -539,6 +540,7 @@ func (dm *KubeArmorDaemon) UpdateContainerdContainer(ctx context.Context, contai
 				}
 				idx++
 			}
+			dm.EndPointsLock.Unlock()
 		}
 
 		dm.EndPointsLock.Lock()

--- a/KubeArmor/core/kubeUpdate.go
+++ b/KubeArmor/core/kubeUpdate.go
@@ -1102,15 +1102,14 @@ func containsPolicy(endPointPolicies []tp.SecurityPolicy, secPolicy tp.SecurityP
 
 // UpdateSecurityPolicy Function
 func (dm *KubeArmorDaemon) UpdateSecurityPolicy(action string, secPolicyType string, secPolicy tp.SecurityPolicy) {
+	// Copy endpoints under lock to prevent TOCTOU race condition
+	// where endpoint array could shrink between length check and access
 	dm.EndPointsLock.RLock()
-	endPointsLength := len(dm.EndPoints)
+	endPointsCopy := make([]tp.EndPoint, len(dm.EndPoints))
+	copy(endPointsCopy, dm.EndPoints)
 	dm.EndPointsLock.RUnlock()
 
-	for idx := range endPointsLength {
-		dm.EndPointsLock.RLock()
-		endPoint := dm.EndPoints[idx]
-		dm.EndPointsLock.RUnlock()
-
+	for _, endPoint := range endPointsCopy {
 		// update a security policy
 		if secPolicyType == KubeArmorPolicy {
 			if len(secPolicy.Spec.Selector.Containers) == 0 || kl.ContainsElement(secPolicy.Spec.Selector.Containers, endPoint.ContainerName) {
@@ -1157,30 +1156,37 @@ func (dm *KubeArmorDaemon) UpdateSecurityPolicy(action string, secPolicyType str
 					}
 				}
 
+				// Find and update the original endpoint in the array by matching unique identifiers
 				dm.EndPointsLock.Lock()
-				dm.EndPoints[idx] = endPoint
-				dm.EndPointsLock.Unlock()
+				for idx := range dm.EndPoints {
+					if dm.EndPoints[idx].NamespaceName == endPoint.NamespaceName &&
+						dm.EndPoints[idx].EndPointName == endPoint.EndPointName &&
+						dm.EndPoints[idx].ContainerName == endPoint.ContainerName {
+						dm.EndPoints[idx] = endPoint
 
-				if cfg.GlobalCfg.Policy {
-					// update security policies
-					dm.Logger.UpdateSecurityPolicies("UPDATED", endPoint)
+						if cfg.GlobalCfg.Policy {
+							// update security policies
+							dm.Logger.UpdateSecurityPolicies("UPDATED", endPoint)
 
-					if dm.EndPoints[idx].PolicyEnabled == tp.KubeArmorPolicyEnabled {
-						// enforce security policies
-						if !kl.ContainsElement(cfg.GlobalCfg.ConfigUntrackedNs.Load().([]string), dm.EndPoints[idx].NamespaceName) || action == deleteEvent {
-							// we want to avoid new policies in untracked namespaces but deletion of the existing policies should be allowed
-							if dm.RuntimeEnforcer != nil {
-								dm.RuntimeEnforcer.UpdateSecurityPolicies(dm.EndPoints[idx])
+							if dm.EndPoints[idx].PolicyEnabled == tp.KubeArmorPolicyEnabled {
+								// enforce security policies
+								if !kl.ContainsElement(cfg.GlobalCfg.ConfigUntrackedNs.Load().([]string), dm.EndPoints[idx].NamespaceName) || action == deleteEvent {
+									// we want to avoid new policies in untracked namespaces but deletion of the existing policies should be allowed
+									if dm.RuntimeEnforcer != nil {
+										dm.RuntimeEnforcer.UpdateSecurityPolicies(dm.EndPoints[idx])
+									}
+									if dm.Presets != nil {
+										dm.Presets.UpdateSecurityPolicies(dm.EndPoints[idx])
+									}
+								} else {
+									dm.Logger.Warnf("Policy cannot be enforced in untracked namespace %s", dm.EndPoints[idx].NamespaceName)
+								}
 							}
-							if dm.Presets != nil {
-								dm.Presets.UpdateSecurityPolicies(dm.EndPoints[idx])
-							}
-						} else {
-							dm.Logger.Warnf("Policy cannot be enforced in untracked namespace %s", dm.EndPoints[idx].NamespaceName)
 						}
+						break
 					}
-
 				}
+				dm.EndPointsLock.Unlock()
 			}
 		} else if secPolicyType == KubeArmorClusterPolicy {
 			// additional OR check added with containsPolicy() is when this endPoint's ns is removed from secPolicy.Spec.Selector.MatchExpressions[i].Values
@@ -1231,29 +1237,37 @@ func (dm *KubeArmorDaemon) UpdateSecurityPolicy(action string, secPolicyType str
 					}
 				}
 
+				// Find and update the original endpoint in the array by matching unique identifiers
 				dm.EndPointsLock.Lock()
-				dm.EndPoints[idx] = endPoint
-				dm.EndPointsLock.Unlock()
+				for idx := range dm.EndPoints {
+					if dm.EndPoints[idx].NamespaceName == endPoint.NamespaceName &&
+						dm.EndPoints[idx].EndPointName == endPoint.EndPointName &&
+						dm.EndPoints[idx].ContainerName == endPoint.ContainerName {
+						dm.EndPoints[idx] = endPoint
 
-				if cfg.GlobalCfg.Policy {
-					// update security policies
-					dm.Logger.UpdateSecurityPolicies("UPDATED", endPoint)
+						if cfg.GlobalCfg.Policy {
+							// update security policies
+							dm.Logger.UpdateSecurityPolicies("UPDATED", endPoint)
 
-					if dm.EndPoints[idx].PolicyEnabled == tp.KubeArmorPolicyEnabled {
-						// enforce security policies
-						if !kl.ContainsElement(cfg.GlobalCfg.ConfigUntrackedNs.Load().([]string), dm.EndPoints[idx].NamespaceName) || action == deleteEvent {
-							// we want to avoid new policies in untracked namespaces but deletion of the existing policies should be allowed
-							if dm.RuntimeEnforcer != nil {
-								dm.RuntimeEnforcer.UpdateSecurityPolicies(dm.EndPoints[idx])
+							if dm.EndPoints[idx].PolicyEnabled == tp.KubeArmorPolicyEnabled {
+								// enforce security policies
+								if !kl.ContainsElement(cfg.GlobalCfg.ConfigUntrackedNs.Load().([]string), dm.EndPoints[idx].NamespaceName) || action == deleteEvent {
+									// we want to avoid new policies in untracked namespaces but deletion of the existing policies should be allowed
+									if dm.RuntimeEnforcer != nil {
+										dm.RuntimeEnforcer.UpdateSecurityPolicies(dm.EndPoints[idx])
+									}
+									if dm.Presets != nil {
+										dm.Presets.UpdateSecurityPolicies(dm.EndPoints[idx])
+									}
+								} else {
+									dm.Logger.Warnf("Policy cannot be enforced in untracked namespace %s", dm.EndPoints[idx].NamespaceName)
+								}
 							}
-							if dm.Presets != nil {
-								dm.Presets.UpdateSecurityPolicies(dm.EndPoints[idx])
-							}
-						} else {
-							dm.Logger.Warnf("Policy cannot be enforced in untracked namespace %s", dm.EndPoints[idx].NamespaceName)
 						}
+						break
 					}
 				}
+				dm.EndPointsLock.Unlock()
 			}
 		}
 	}
@@ -1950,17 +1964,15 @@ func (dm *KubeArmorDaemon) WatchClusterSecurityPolicies(timeout time.Duration) c
 
 // UpdateHostSecurityPolicies Function
 func (dm *KubeArmorDaemon) UpdateHostSecurityPolicies() {
+	// Copy host security policies under lock to prevent TOCTOU race condition
 	dm.HostSecurityPoliciesLock.RLock()
-	hostSecurityPoliciesLength := len(dm.HostSecurityPolicies)
+	hostSecurityPoliciesCopy := make([]tp.HostSecurityPolicy, len(dm.HostSecurityPolicies))
+	copy(hostSecurityPoliciesCopy, dm.HostSecurityPolicies)
 	dm.HostSecurityPoliciesLock.RUnlock()
 
 	secPolicies := []tp.HostSecurityPolicy{}
 
-	for idx := range hostSecurityPoliciesLength {
-		dm.EndPointsLock.RLock()
-		policy := dm.HostSecurityPolicies[idx]
-		dm.EndPointsLock.RUnlock()
-
+	for _, policy := range hostSecurityPoliciesCopy {
 		if kl.MatchIdentities(policy.Spec.NodeSelector.Identities, dm.Node.Identities) {
 			secPolicies = append(secPolicies, policy)
 		}
@@ -2662,14 +2674,13 @@ func (dm *KubeArmorDaemon) UpdateDefaultPosture(action string, namespace string,
 	}
 	dm.Logger.UpdateDefaultPosture(action, namespace, defaultPosture)
 
+	// Copy endpoints under lock to prevent TOCTOU race condition
 	dm.EndPointsLock.RLock()
-	endPointsLen := len(dm.EndPoints)
+	endPointsCopy := make([]tp.EndPoint, len(dm.EndPoints))
+	copy(endPointsCopy, dm.EndPoints)
 	dm.EndPointsLock.RUnlock()
 
-	for idx := range endPointsLen {
-		dm.EndPointsLock.RLock()
-		endPoint := dm.EndPoints[idx]
-		dm.EndPointsLock.RUnlock()
+	for _, endPoint := range endPointsCopy {
 		// update a security policy
 		if namespace == endPoint.NamespaceName {
 			if endPoint.DefaultPosture == defaultPosture {
@@ -2679,8 +2690,16 @@ func (dm *KubeArmorDaemon) UpdateDefaultPosture(action string, namespace string,
 			dm.Logger.Printf("Updating default posture for %s with %v namespace default %v", endPoint.EndPointName, endPoint.DefaultPosture, defaultPosture)
 			endPoint.DefaultPosture = defaultPosture
 
+			// Find and update the original endpoint in the array by matching unique identifiers
 			dm.EndPointsLock.Lock()
-			dm.EndPoints[idx] = endPoint
+			for idx := range dm.EndPoints {
+				if dm.EndPoints[idx].NamespaceName == endPoint.NamespaceName &&
+					dm.EndPoints[idx].EndPointName == endPoint.EndPointName &&
+					dm.EndPoints[idx].ContainerName == endPoint.ContainerName {
+					dm.EndPoints[idx] = endPoint
+					break
+				}
+			}
 			dm.EndPointsLock.Unlock()
 
 			if cfg.GlobalCfg.Policy {

--- a/KubeArmor/core/kubeUpdate_race_test.go
+++ b/KubeArmor/core/kubeUpdate_race_test.go
@@ -1,0 +1,324 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 Authors of KubeArmor
+
+package core
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	tp "github.com/kubearmor/KubeArmor/KubeArmor/types"
+)
+
+// TestUpdateSecurityPolicyRaceCondition tests that UpdateSecurityPolicy doesn't panic
+// when endpoints are concurrently added/removed during policy updates
+func TestUpdateSecurityPolicyRaceCondition(t *testing.T) {
+	// Create a test daemon
+	dm := &KubeArmorDaemon{
+		EndPoints:                 []tp.EndPoint{},
+		EndPointsLock:             &sync.RWMutex{},
+		SecurityPolicies:          []tp.SecurityPolicy{},
+		SecurityPoliciesLock:      &sync.RWMutex{},
+		HostSecurityPolicies:      []tp.HostSecurityPolicy{},
+		HostSecurityPoliciesLock:  &sync.RWMutex{},
+		DefaultPostures:           map[string]tp.DefaultPosture{},
+		DefaultPosturesLock:       &sync.Mutex{},
+	}
+
+	// Initialize with some test endpoints
+	for i := 0; i < 50; i++ {
+		endpoint := tp.EndPoint{
+			NamespaceName:    "test-namespace",
+			EndPointName:     fmt.Sprintf("test-endpoint-%d", i),
+			ContainerName:    fmt.Sprintf("test-container-%d", i),
+			Identities:       []string{fmt.Sprintf("label=value-%d", i)},
+			SecurityPolicies: []tp.SecurityPolicy{},
+		}
+		dm.EndPoints = append(dm.EndPoints, endpoint)
+	}
+
+	// Test security policy
+	testPolicy := tp.SecurityPolicy{
+		Metadata: map[string]string{
+			"namespaceName": "test-namespace",
+			"policyName":    "test-policy",
+		},
+		Spec: tp.SecuritySpec{
+			Selector: tp.SelectorType{
+				Identities: []string{"label=value-1"},
+			},
+		},
+	}
+
+	// Run concurrent operations
+	var wg sync.WaitGroup
+	stopCh := make(chan struct{})
+
+	// Goroutine 1: Constantly update security policies
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				// This should not panic even if endpoints array changes
+				dm.UpdateSecurityPolicy("add", "KubeArmorPolicy", testPolicy)
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	}()
+
+	// Goroutine 2: Constantly add/remove endpoints
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		counter := 100
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				dm.EndPointsLock.Lock()
+				if len(dm.EndPoints) > 0 {
+					// Remove an endpoint
+					dm.EndPoints = dm.EndPoints[:len(dm.EndPoints)-1]
+				}
+				// Add a new endpoint
+				newEndpoint := tp.EndPoint{
+					NamespaceName:    "test-namespace",
+					EndPointName:     fmt.Sprintf("dynamic-endpoint-%d", counter),
+					ContainerName:    fmt.Sprintf("dynamic-container-%d", counter),
+					Identities:       []string{fmt.Sprintf("label=value-%d", counter)},
+					SecurityPolicies: []tp.SecurityPolicy{},
+				}
+				dm.EndPoints = append(dm.EndPoints, newEndpoint)
+				counter++
+				dm.EndPointsLock.Unlock()
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	}()
+
+	// Let the race condition test run for a short time
+	time.Sleep(100 * time.Millisecond)
+	close(stopCh)
+	wg.Wait()
+
+	t.Log("UpdateSecurityPolicy race condition test completed without panics")
+}
+
+// TestUpdateHostSecurityPoliciesRaceCondition tests race conditions in UpdateHostSecurityPolicies
+func TestUpdateHostSecurityPoliciesRaceCondition(t *testing.T) {
+	dm := &KubeArmorDaemon{
+		HostSecurityPolicies:     []tp.HostSecurityPolicy{},
+		HostSecurityPoliciesLock: &sync.RWMutex{},
+		Node: tp.Node{
+			Identities: []string{"hostname=test-node"},
+		},
+	}
+
+	// Initialize with test host policies
+	for i := 0; i < 20; i++ {
+		policy := tp.HostSecurityPolicy{
+			Metadata: map[string]string{
+				"policyName": fmt.Sprintf("host-policy-%d", i),
+			},
+			Spec: tp.HostSecuritySpec{
+				NodeSelector: tp.NodeSelectorType{
+					Identities: []string{"hostname=test-node"},
+				},
+			},
+		}
+		dm.HostSecurityPolicies = append(dm.HostSecurityPolicies, policy)
+	}
+
+	var wg sync.WaitGroup
+	stopCh := make(chan struct{})
+
+	// Goroutine 1: Update host security policies
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				dm.UpdateHostSecurityPolicies()
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	}()
+
+	// Goroutine 2: Modify host policies
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		counter := 100
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				dm.HostSecurityPoliciesLock.Lock()
+				if len(dm.HostSecurityPolicies) > 0 {
+					// Remove a policy
+					dm.HostSecurityPolicies = dm.HostSecurityPolicies[:len(dm.HostSecurityPolicies)-1]
+				}
+				// Add a new policy
+				newPolicy := tp.HostSecurityPolicy{
+					Metadata: map[string]string{
+						"policyName": fmt.Sprintf("dynamic-host-policy-%d", counter),
+					},
+					Spec: tp.HostSecuritySpec{
+						NodeSelector: tp.NodeSelectorType{
+							Identities: []string{"hostname=test-node"},
+						},
+					},
+				}
+				dm.HostSecurityPolicies = append(dm.HostSecurityPolicies, newPolicy)
+				counter++
+				dm.HostSecurityPoliciesLock.Unlock()
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	}()
+
+	// Run test
+	time.Sleep(50 * time.Millisecond)
+	close(stopCh)
+	wg.Wait()
+
+	t.Log("UpdateHostSecurityPolicies race condition test completed without panics")
+}
+
+// TestUpdateDefaultPostureRaceCondition tests race conditions in UpdateDefaultPosture
+func TestUpdateDefaultPostureRaceCondition(t *testing.T) {
+	// Skip this test for now due to external dependencies (Logger)
+	t.Skip("Skipping UpdateDefaultPosture test due to external dependencies")
+	
+	dm := &KubeArmorDaemon{
+		EndPoints:               []tp.EndPoint{},
+		EndPointsLock:           &sync.RWMutex{},
+		DefaultPostures:         map[string]tp.DefaultPosture{},
+		DefaultPosturesLock:     &sync.Mutex{},
+	}
+
+	// Initialize with test endpoints
+	for i := 0; i < 30; i++ {
+		endpoint := tp.EndPoint{
+			NamespaceName:   "test-namespace",
+			EndPointName:    fmt.Sprintf("test-endpoint-%d", i),
+			ContainerName:   fmt.Sprintf("test-container-%d", i),
+			DefaultPosture:  tp.DefaultPosture{FileAction: "allow"},
+		}
+		dm.EndPoints = append(dm.EndPoints, endpoint)
+	}
+
+	var wg sync.WaitGroup
+	stopCh := make(chan struct{})
+
+	// Goroutine 1: Update default postures
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				newPosture := tp.DefaultPosture{FileAction: "block"}
+				dm.UpdateDefaultPosture("add", "test-namespace", newPosture, true)
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	}()
+
+	// Goroutine 2: Modify endpoints
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		counter := 100
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				dm.EndPointsLock.Lock()
+				if len(dm.EndPoints) > 0 {
+					// Remove an endpoint
+					dm.EndPoints = dm.EndPoints[:len(dm.EndPoints)-1]
+				}
+				// Add a new endpoint
+				newEndpoint := tp.EndPoint{
+					NamespaceName:  "test-namespace",
+					EndPointName:   fmt.Sprintf("dynamic-endpoint-%d", counter),
+					ContainerName:  fmt.Sprintf("dynamic-container-%d", counter),
+					DefaultPosture: tp.DefaultPosture{FileAction: "allow"},
+				}
+				dm.EndPoints = append(dm.EndPoints, newEndpoint)
+				counter++
+				dm.EndPointsLock.Unlock()
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+	}()
+
+	// Run test
+	time.Sleep(50 * time.Millisecond)
+	close(stopCh)
+	wg.Wait()
+
+	t.Log("UpdateDefaultPosture race condition test completed without panics")
+}
+
+// TestConcurrentEndpointAccess tests general concurrent access patterns
+func TestConcurrentEndpointAccess(t *testing.T) {
+	dm := &KubeArmorDaemon{
+		EndPoints:     []tp.EndPoint{},
+		EndPointsLock: &sync.RWMutex{},
+	}
+
+	var wg sync.WaitGroup
+	numGoroutines := 10
+	operations := 1000
+
+	// Create multiple goroutines that concurrently access endpoints
+	for i := 0; i < numGoroutines; i++ {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+			for j := 0; j < operations; j++ {
+				// Read operation - copy endpoints safely
+				dm.EndPointsLock.RLock()
+				endPointsCopy := make([]tp.EndPoint, len(dm.EndPoints))
+				copy(endPointsCopy, dm.EndPoints)
+				dm.EndPointsLock.RUnlock()
+
+				// Process the copy (simulate work)
+				for _, ep := range endPointsCopy {
+					_ = ep.NamespaceName // Do something with the endpoint
+				}
+
+				// Write operation - modify endpoints
+				if j%10 == 0 { // Every 10th operation is a write
+					dm.EndPointsLock.Lock()
+					newEndpoint := tp.EndPoint{
+						NamespaceName: fmt.Sprintf("ns-%d", id),
+						EndPointName:  fmt.Sprintf("ep-%d-%d", id, j),
+						ContainerName: fmt.Sprintf("container-%d-%d", id, j),
+					}
+					dm.EndPoints = append(dm.EndPoints, newEndpoint)
+					dm.EndPointsLock.Unlock()
+				}
+			}
+		}(i)
+	}
+
+	wg.Wait()
+	t.Logf("Concurrent access test completed. Final endpoint count: %d", len(dm.EndPoints))
+}

--- a/KubeArmor/core/kubeUpdate_race_test.go
+++ b/KubeArmor/core/kubeUpdate_race_test.go
@@ -1,10 +1,11 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright 2024 Authors of KubeArmor
+// Copyright 2026 Authors of KubeArmor
 
 package core
 
 import (
 	"fmt"
+	"math/rand"
 	"sync"
 	"testing"
 	"time"
@@ -17,14 +18,14 @@ import (
 func TestUpdateSecurityPolicyRaceCondition(t *testing.T) {
 	// Create a test daemon
 	dm := &KubeArmorDaemon{
-		EndPoints:                 []tp.EndPoint{},
-		EndPointsLock:             &sync.RWMutex{},
-		SecurityPolicies:          []tp.SecurityPolicy{},
-		SecurityPoliciesLock:      &sync.RWMutex{},
-		HostSecurityPolicies:      []tp.HostSecurityPolicy{},
-		HostSecurityPoliciesLock:  &sync.RWMutex{},
-		DefaultPostures:           map[string]tp.DefaultPosture{},
-		DefaultPosturesLock:       &sync.Mutex{},
+		EndPoints:                []tp.EndPoint{},
+		EndPointsLock:            &sync.RWMutex{},
+		SecurityPolicies:         []tp.SecurityPolicy{},
+		SecurityPoliciesLock:     &sync.RWMutex{},
+		HostSecurityPolicies:     []tp.HostSecurityPolicy{},
+		HostSecurityPoliciesLock: &sync.RWMutex{},
+		DefaultPostures:          map[string]tp.DefaultPosture{},
+		DefaultPosturesLock:      &sync.Mutex{},
 	}
 
 	// Initialize with some test endpoints
@@ -85,7 +86,9 @@ func TestUpdateSecurityPolicyRaceCondition(t *testing.T) {
 				dm.EndPointsLock.Lock()
 				if len(dm.EndPoints) > 0 {
 					// Remove an endpoint
-					dm.EndPoints = dm.EndPoints[:len(dm.EndPoints)-1]
+					// dm.EndPoints = dm.EndPoints[:len(dm.EndPoints)-1]
+					shrink := rand.Intn(len(dm.EndPoints))
+					dm.EndPoints = dm.EndPoints[:len(dm.EndPoints)-shrink]
 				}
 				// Add a new endpoint
 				newEndpoint := tp.EndPoint{
@@ -98,13 +101,12 @@ func TestUpdateSecurityPolicyRaceCondition(t *testing.T) {
 				dm.EndPoints = append(dm.EndPoints, newEndpoint)
 				counter++
 				dm.EndPointsLock.Unlock()
-				time.Sleep(1 * time.Millisecond)
 			}
 		}
 	}()
 
 	// Let the race condition test run for a short time
-	time.Sleep(100 * time.Millisecond)
+	time.Sleep(50 * time.Millisecond)
 	close(stopCh)
 	wg.Wait()
 
@@ -200,21 +202,21 @@ func TestUpdateHostSecurityPoliciesRaceCondition(t *testing.T) {
 func TestUpdateDefaultPostureRaceCondition(t *testing.T) {
 	// Skip this test for now due to external dependencies (Logger)
 	t.Skip("Skipping UpdateDefaultPosture test due to external dependencies")
-	
+
 	dm := &KubeArmorDaemon{
-		EndPoints:               []tp.EndPoint{},
-		EndPointsLock:           &sync.RWMutex{},
-		DefaultPostures:         map[string]tp.DefaultPosture{},
-		DefaultPosturesLock:     &sync.Mutex{},
+		EndPoints:           []tp.EndPoint{},
+		EndPointsLock:       &sync.RWMutex{},
+		DefaultPostures:     map[string]tp.DefaultPosture{},
+		DefaultPosturesLock: &sync.Mutex{},
 	}
 
 	// Initialize with test endpoints
 	for i := 0; i < 30; i++ {
 		endpoint := tp.EndPoint{
-			NamespaceName:   "test-namespace",
-			EndPointName:    fmt.Sprintf("test-endpoint-%d", i),
-			ContainerName:   fmt.Sprintf("test-container-%d", i),
-			DefaultPosture:  tp.DefaultPosture{FileAction: "allow"},
+			NamespaceName:  "test-namespace",
+			EndPointName:   fmt.Sprintf("test-endpoint-%d", i),
+			ContainerName:  fmt.Sprintf("test-container-%d", i),
+			DefaultPosture: tp.DefaultPosture{FileAction: "allow"},
 		}
 		dm.EndPoints = append(dm.EndPoints, endpoint)
 	}

